### DISCUSSION TITLE
Request forwarding with large files

### DISCUSSION THREAD

**Discussion by thomafred** [Neoteroi/BlackSheep#350] [Q&A] [answered]
I am working on a reverse-proxy-like thing where a user uploads a file to a central endpoint, which in turn forwards the file to an appropriate micro-service in the backend.

The reverse-proxy works pretty well in its current state, however I have concerns related to handling large files, especially related to handling multiple simultaneous requests that may include large files.

How do I ensure that the reverse-proxy is able to process multiple simultaneous requests without blowing up the memory-footprint of the app?

**Comment by RobertoPrevato** [comment]
Hi @thomafred 
Interesting question. I would start by ensuring that files are processed in chunks, meaning:
- the server reads the files in chunks
- the server forwards the files chunk by chunk, using the functions of the library used for the HTTP client
- your micro-service is also able to handle files in chunks

This way, files are **never** read whole in memory at once.

I do NOT recommend using multipart/form-data for this kind of scenario, although it is the default used by HTML <forms> having `file` inputs, because it makes the API more complicated.

To read the input in chunks in blacksheep (server part), you can do:

```python
from blacksheep import Application, Request, Response

app = Application()
post = app.router.post

@post("/upload")
async def save_big_file(request: Request):

    async for chunk in request.stream():
        # TODO: do something with the chunk, in this case forward the chunk to a back-end
        ...

    # TODO: send a request to the back-end, return the desired response
    return Response(200)


if __name__ == "__main__":
    import uvicorn

    uvicorn.run(app, port=44555)

```

This is documented [here](https://www.neoteroi.dev/blacksheep/requests/#reading-streams).

Are you using the BlackSheep HTTP client or another client? If you are using the client in BlackSheep, I can prepare an example for uploading files in chunks, too.

Processing the files in chunks ensures that things are done efficiently. Then I would test with the expected workload - if things work, fine, otherwise I would consider two options:
- one is to increase the resources of the machine where the reverse-proxy runs, if possible
- second is to implement a rate limiter. There is no built-in support for this in blacksheep, though

As a side note, this depends on how your micro-service is implemented. For example, the Azure Storage Blob API offers two operations: one to upload a single chunk, one to communicate the end of the upload process. That approach is straightforward, and it forces clients to sends files in exact chunks, with multiple web requests.

**Reply by thomafred** [reply]
@RobertoPrevato I am using the blacksheep client for this.

Based on your input I have cobbled together the following example:

```python
# Request forwarder
def strip_headers(request):
    headers = {}
    for header, value in request.headers.items():
        lheader = header.lower()

        if lheader in (
                b"content-length",
                b"transfer-encoding",
                b"content-encoding",
                b"connection",
                b"host"
        ):
            continue  # Ignore

    return [(key, value) for key, value in headers.items()]

@post("/upload")
async def forward_big_file(request: Request):

    async def data_generator():
        async for chunk in request.stream():
            yield chunk

    # TODO: Handle params

    content = StreamedContent(request.content_type(), data_generator)
    req = Request(request.method, b"http://localhost:44556/upload", strip_headers(request)).with_content(content)

    async with ClientSession() as client:
        response = await client.send(req)

    # TODO: send a request to the back-end, return the desired response
    return Response(response.status)


if __name__ == "__main__":
    import uvicorn

    uvicorn.run(app, port=44555)
```

```python
# File request handler

from blacksheep import Application, Request, Response

app = Application()
post = app.router.post

@post("/upload")
async def save_big_file(request: Request):

    bytes_received = 0
    async for chunk in request.stream():
        bytes_received += len(chunk)

    print(f"Received {bytes_received} bytes")

    # TODO: send a request to the back-end, return the desired response
    return Response(200)


if __name__ == "__main__":
    import uvicorn

    uvicorn.run(app, port=44556)
```

The request-forwarder will strip the header values for `content-length`, `transfer-encoding`, `content-encoding`, `connection` and `host`, so not to interfere with the request forwarding.

I am using the following command to test:

```shell
curl --request POST "localhost:44555/upload" --data-binary @./myfile.bin
```

This works fine when uploading the file directly to the file request handler, but fails when using the request-forwarder. From the log, I can see the following warning:

```
WARNING:  Invalid HTTP request received.
```

From what I can tell, the warning is not originating from blacksheep, but possibly from uvicorn.

I was able to reproduce this using Postman setting the `transfer-encoding=chunked` when making request directly to the file request handler. When disabled the `content-length` header in Postman, everything worked okay.

**Comment by thomafred** [comment]
Hi @RobertoPrevato. Hope you are doing well :)

I have had a second look at this, and I have been unable to resolve the `Invalid HTTP request received`-error.

Do you have any ideas on how we might work around this?

**Comment by RobertoPrevato** [comment]
Hi @thomafred 
I am doing well, thank You! I hope the same holds true for you. 😄 
Sorry for not investigating earlier. I am investigating today, and I created an example from scratch (to not be influenced by yours).

I can reproduce the error you reported about the content-length. I did not find the root cause of the problem, yet, but I think to be on a good track.

PS. lately I might have seemed inactive on GitHub. I had less time (I was busier than usual), and I am working on something privately, still (a CLI to bootstrap BlackSheep projects).

**Comment by RobertoPrevato** [comment, marked as answer]
Hi @thomafred 
I prepared a working example, and published it here: https://github.com/Neoteroi/BlackSheep-Examples/tree/main/proxy

The BlackSheep server in my example handles files of any size properly (lazily, in chunks) - it is the proxy part.
Please take a look at how I implemented it, and the notes I left there.

The example is not 100% complete, as it should handle better some cases, but it describes what you asked in this discussion.

**Comment by thomafred** [comment]
Hi @RobertoPrevato,

This is awesome, thank you so much!

Based on your example, I can see the request coming in the backend service and I can see the backend service response being returned (A `200`-status with a json-body), but the proxy is still receiving a `400 - Invalid HTTP request received.`.

Are there any special considerations we need to do in the backend service too?

**Reply by RobertoPrevato** [reply]
@thomafred thanks for your patience! I will check more, and using your example, too. By the way, while preparing the example I realized that maybe BlackSheep should support async generators to prepare chunked responses while **not** using 'Transfer-Encoding: chunked'.

**Reply by thomafred** [reply]
> By the way, while preparing the example I realized that maybe BlackSheep should support async generators to prepare chunked responses while not using 'Transfer-Encoding: chunked'.

Yes, I agree this might help in this case.

**Reply by RobertoPrevato** [reply]
Hi @thomafred 
I tried creating a proxy example between two BlackSheep applications (proxy A and back-end B), and even though I managed to proxy properly the requests from A to B, and B handles the request as desired when uploading files with multipart/form-data, the **uvicorn** server of B `httptools` part raises a cryptic exception at the end of the request-response cycle and finally returns a Bad Request response to the proxy server A (the same [reported here](https://github.com/encode/uvicorn/issues/344) and several other places).

I will try again using `h11` implementation for `B` and leave a new comment under this one, to share the result.

**Reply by RobertoPrevato** [reply]
Running the `uvicorn` server using `--http h11` or `http="h11"` helps in my case, and I get the proper response between the proxy and the back-end. For some reason, the back-end still issues a warning for reasons I do not understand: "WARNING:  Invalid HTTP request received.".

I do not understand why this is happening because the Flask server in my first example does not complain about anything, and everything else works. I will try to look more into this, but it might be an issue in `uvicorn` itself.

**Comment by RobertoPrevato** [comment]
Hi @thomafred 
I finally investigated more and detected the issues. I created a new release for `v1.2.16` to correct problems that would make it difficult to create a fully functional HTTP Proxy. Tomorrow I will publish an updated example for the HTTP Proxy, that can be much better now, thanks to the fixed introduced with 1.2.16.

**Reply by thomafred** [reply]
This is amazing, thank you so much!

I will make another attempt in our proxy implementation and let you know how it goes :)

**Reply by thomafred** [reply]
This worked like a charm!

Once again - thank you so much :rocket:

**Reply by RobertoPrevato** [reply]
Thank You @thomafred, hadn't you created this discussion, I would have not noticed there was something wrong. :)

I published the updated examples here:
- [proxy-1](https://github.com/Neoteroi/BlackSheep-Examples/tree/main/proxy-1)
- [proxy-2](https://github.com/Neoteroi/BlackSheep-Examples/tree/main/proxy-2)

All issues were boiling down to the fact that BlackSheep supported only `transfer-encoding: chunked` when using asynchronous generators to build request / response content. Now it also supports async generators when the content-length is known. It is much easier to implement an HTTP Proxy (otherwise one would have to parse the content sent without transfer-encoding and alter it to be transfer-encoding compatible! I also did not realize this immediately).

**Comment by thomafred** [comment]
Hi @RobertoPrevato 

Sorry to be pulling at this thread. I am seeing some issues related to small bodies with the `StreamedContent`:

```
  File "/home/tom/.cache/pypoetry/virtualenvs/app-odcat-Dj0wURcA-py3.10/lib64/python3.10/site-packages/blacksheep/client/session.py", line 371, in send
    response = await self._handler(request)
  File "/home/tom/.cache/pypoetry/virtualenvs/app-odcat-Dj0wURcA-py3.10/lib64/python3.10/site-packages/blacksheep/middlewares.py", line 6, in middleware_wrapper
    return await handler(request, next_handler)
  File "/home/tom/.cache/pypoetry/virtualenvs/app-odcat-Dj0wURcA-py3.10/lib64/python3.10/site-packages/blacksheep/client/cookies.py", line 316, in cookies_middleware
    response = await next_handler(request)
  File "/home/tom/.cache/pypoetry/virtualenvs/app-odcat-Dj0wURcA-py3.10/lib64/python3.10/site-packages/blacksheep/client/session.py", line 211, in root_handler
    return await self._send_core(request)
  File "/home/tom/.cache/pypoetry/virtualenvs/app-odcat-Dj0wURcA-py3.10/lib64/python3.10/site-packages/blacksheep/client/session.py", line 390, in _send_core
    return await self._send_using_connection(request)
  File "/home/tom/.cache/pypoetry/virtualenvs/app-odcat-Dj0wURcA-py3.10/lib64/python3.10/site-packages/blacksheep/client/session.py", line 396, in _send_using_connection
    return await asyncio.wait_for(
  File "/usr/lib64/python3.10/asyncio/tasks.py", line 445, in wait_for
    return fut.result()
  File "/home/tom/.cache/pypoetry/virtualenvs/app-odcat-Dj0wURcA-py3.10/lib64/python3.10/site-packages/blacksheep/client/connection.py", line 279, in send
    self.transport.write(write_small_request(request))
  File "blacksheep/scribe.pyx", line 159, in blacksheep.scribe.write_small_request
  File "blacksheep/scribe.pyx", line 169, in blacksheep.scribe.write_small_request
TypeError: can't extend bytearray with NoneType
INFO:     127.0.0.1:42392 - "POST /resourceStore/list HTTP/1.1" 500 Internal Server Error
```

Setting a breakpoint at [`blacksheep/client/connection.py:279`](https://github.com/Neoteroi/BlackSheep/blob/v1.2.16/blacksheep/client/connection.py#LL279C19-L279C19), it seems the content body is `None`.

This does not appear to be supported by [`write_small_requests`](https://github.com/Neoteroi/BlackSheep/blob/v1.2.16/blacksheep/scribe.pyx#LL169C42-L169C42).

Should the body be populated when dealing small requests?

**Reply by RobertoPrevato** [reply]
@thomafred sorry for that mistake. I just released `1.2.17` to address that. It was a regression because earlier `StreamedContent` would always have content-length == -1 and would result in chunked transfer encoding being used.
I added more tests and I think now it works as it should. Please upgrade to `1.2.17` and let me know if you encounter more issues.

**Reply by thomafred** [reply]
@RobertoPrevato Thank you! This seems to have fixed the small requests. However not providing the `content-length` in the `StreamedContent`, still gives the `Invalid HTTP Request`-error.

Another observation: The way my API is architected, the requests bounce around a bit inside the API microservices before a response is returned to the users. For example:

![image](https://github.com/Neoteroi/BlackSheep/assets/7789756/1f144941-9fab-44c2-9b30-be3267f9e39a)

I am noticing that the request is successfully forwarded from the reverse proxy the first time, so that it is successfully received and handled by the microservice. Next, the microservice retrieves some state based on the user request from a state-controller (a wrapper around a database). The state-request also needs to cross the proxy as it is intended to authorize this request. When the state-request reaches the state-controller, the `Invalid HTTP Request` is once again returned.

I am able to make requests directly to the state-controller, but not have requests from the microservice for some reason.

Both the request to the microservice and to the state controller is quite small (small JSON < 1kB).

**Reply by RobertoPrevato** [reply]
You're welcome, @thomafred. You really need to provide the content-length if the client provides it, like in the examples I published in the `BlackSheep-Examples` repository. This is because, if you don't provide any specific Content-Lenght, BlackSheep will default to using ["Transfer-Encoding: chunked"](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Transfer-Encoding), which is a good default when you don't know the content-length _a priori_ and you use an async generator.

However, `Transfer-Encoding: chunked` means that you are sending chunks to the clients in a particular way, communicating the length of each chunk to the client. If you simply proxy the request content stream, you need to respect what the original client is saying (you need to proxy the right content-length).

> Data is sent in a series of chunks. The [Content-Length](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Length) header is omitted in this case and at the beginning of each chunk you need to add the length of the current chunk in hexadecimal format, followed by '\r\n' and then the chunk itself, followed by another '\r\n'. The terminating chunk is a regular chunk, with the exception that its length is zero. It is followed by the trailer, which consists of a (possibly empty) sequence of header fields.

~~Now, what happens if you don't proxy the `Content-Length` exactly (like in my example) and you stream the request body into the response body (without parsing it all in-memory and re-writing it, which would be memory inefficient), is that you are proxying a request body stream that is not using chunked transfer encoding while communicating to the client of the proxy that it is using chunked encoding (through a response header).~~ < On a second thought, since `uvicorn` handles the chunked encoding by itself, the chunks would be written with their lengths, but when you apply twice you end up with scrambled content.

I think that my example `proxy-2` is replicating a compatible scenario to what you described initially (both proxy and client implemented with blacksheep) and it works. I made several tests with Firefox (initially I only tested uploading big files ☺️).

**Reply by thomafred** [reply]
Cool! I updated our implementation based on your example 2 and things seems to work quite well. Will do some more testing but the level of optimism is quite high :)

**Reply by thomafred** [reply]
This worked really well, but I am encountering an integer overflow error for large files. I have created an issue for this: https://github.com/Neoteroi/BlackSheep/issues/394

We have been using `wod_gld_2013.nc` form [here](https://www.ncei.noaa.gov/thredds-ocean/catalog/ncei/wod/2013/catalog.html) in order to reproduce this issue.

**Reply by thomafred** [reply]
I attempted to update the `content_type` of `StreamContent` from `int` to `cython.long`, which fixed the overflow, but the receiving server is responding with a `Invalid HTTP request received.`-response. Tried with a smaller file, which still works well.

**Reply by thomafred** [reply]
> Running the uvicorn server using --http h11 or http="h11" helps in my case, and I get the proper response between the proxy and the back-end. For some reason, the back-end still issues a warning for reasons I do not understand: "WARNING: Invalid HTTP request received.".

Setting the `http="h11"` setting in uvicorn seems to help in this case too, and we are still getting the `Invalid HTTP request received.`-message, but this time as a warning. The file upload did actually complete.

Exerpt from the uvicorn logs:

```
TRACE:    127.0.0.1:53912 - ASGI [2] Receive {'type': 'http.request', 'body': '<65536 bytes>', 'more_body': True}
TRACE:    127.0.0.1:53912 - ASGI [2] Receive {'type': 'http.request', 'body': '<65536 bytes>', 'more_body': True}
TRACE:    127.0.0.1:53912 - ASGI [2] Receive {'type': 'http.request', 'body': '<65476 bytes>', 'more_body': True}
TRACE:    127.0.0.1:53912 - ASGI [2] Receive {'type': 'http.request', 'body': '<60 bytes>', 'more_body': True}
TRACE:    127.0.0.1:53912 - ASGI [2] Receive {'type': 'http.request', 'body': '<65536 bytes>', 'more_body': True}
TRACE:    127.0.0.1:53912 - ASGI [2] Receive {'type': 'http.request', 'body': '<65476 bytes>', 'more_body': True}
TRACE:    127.0.0.1:53912 - ASGI [2] Receive {'type': 'http.request', 'body': '<60 bytes>', 'more_body': True}
TRACE:    127.0.0.1:53912 - ASGI [2] Receive {'type': 'http.request', 'body': '<65476 bytes>', 'more_body': True}
TRACE:    127.0.0.1:53912 - ASGI [2] Receive {'type': 'http.request', 'body': '<60 bytes>', 'more_body': True}
TRACE:    127.0.0.1:53912 - ASGI [2] Receive {'type': 'http.request', 'body': '<65476 bytes>', 'more_body': True}
TRACE:    127.0.0.1:53912 - ASGI [2] Receive {'type': 'http.request', 'body': '<60 bytes>', 'more_body': True}
TRACE:    127.0.0.1:53912 - ASGI [2] Receive {'type': 'http.request', 'body': '<65536 bytes>', 'more_body': True}
TRACE:    127.0.0.1:53912 - ASGI [2] Receive {'type': 'http.request', 'body': '<65476 bytes>', 'more_body': True}
TRACE:    127.0.0.1:53912 - ASGI [2] Receive {'type': 'http.request', 'body': '<60 bytes>', 'more_body': True}
TRACE:    127.0.0.1:53912 - ASGI [2] Receive {'type': 'http.request', 'body': '<65476 bytes>', 'more_body': True}
TRACE:    127.0.0.1:53912 - ASGI [2] Receive {'type': 'http.request', 'body': '<60 bytes>', 'more_body': True}
TRACE:    127.0.0.1:53912 - ASGI [2] Receive {'type': 'http.request', 'body': '<65476 bytes>', 'more_body': True}
TRACE:    127.0.0.1:53912 - ASGI [2] Receive {'type': 'http.request', 'body': '<60 bytes>', 'more_body': True}
TRACE:    127.0.0.1:53912 - ASGI [2] Receive {'type': 'http.request', 'body': '<65536 bytes>', 'more_body': True}
TRACE:    127.0.0.1:53912 - ASGI [2] Receive {'type': 'http.request', 'body': '<65476 bytes>', 'more_body': True}
TRACE:    127.0.0.1:53912 - ASGI [2] Receive {'type': 'http.request', 'body': '<60 bytes>', 'more_body': True}
TRACE:    127.0.0.1:53912 - ASGI [2] Receive {'type': 'http.request', 'body': '<21989 bytes>', 'more_body': False}
TRACE:httpx._config:load_ssl_context verify=True cert=None trust_env=True http2=False
TRACE:httpx._config:load_verify_locations cafile=/home/tom/.cache/pypoetry/virtualenvs/app-odcat-Dj0wURcA-py3.10/lib/python3.10/site-packages/certifi/cacert.pem
DEBUG:httpx._client:HTTP Request: GET http://127.0.0.1:8888/resourceStore/catalog.hubocean.io/dataset/test-dataset "HTTP/1.1 200 OK"
TRACE:httpx._config:load_ssl_context verify=True cert=None trust_env=True http2=False
TRACE:httpx._config:load_verify_locations cafile=/home/tom/.cache/pypoetry/virtualenvs/app-odcat-Dj0wURcA-py3.10/lib/python3.10/site-packages/certifi/cacert.pem
DEBUG:httpx._client:HTTP Request: POST https://westeurope-1.cognitedata.com/api/v1/projects/oceandata-dev/files/list "HTTP/1.1 200 OK"
DEBUG:odp.odcat.storage.cdffs.async_http_client:status code returned: 200
INFO:odp.odcat.utils.http_exc:HTTP exception caught in request-handler: File name already exists: wod_gld_2013.nc
TRACE:    127.0.0.1:53912 - ASGI [2] Send {'type': 'http.response.start', 'status': 409, 'headers': '<...>'}
INFO:     127.0.0.1:53912 - "PATCH /data/catalog.hubocean.io/dataset/test-dataset/wod_gld_2013.nc HTTP/1.1" 409 Conflict
TRACE:    127.0.0.1:53912 - ASGI [2] Send {'type': 'http.response.body', 'body': '<54 bytes>', 'more_body': False}
WARNING:  Invalid HTTP request received.
TRACE:    127.0.0.1:53912 - ASGI [2] Completed
TRACE:    127.0.0.1:53912 - HTTP connection lost
```

**Reply by thomafred** [reply]
After some more testing, it seems that the forwarded HTTP request is processed and file properly stored. The request handler returns the expected response, but it seems that uvicorn somehow changes to response to the `Invalid HTTP request received.`-message.

The reverse proxy then raises an exception on the response, as the header is indicating a content-length of 28 bytes (my original response), but receives 30 bytes (The `Invalid HTTP request received.`-message length):

```
INFO:     127.0.0.1:57834 - "PATCH /data/catalog.hubocean.io/dataset/test-dataset/wod_gld_2013.nc HTTP/1.1" 400 Bad Request
ERROR:    Exception in ASGI application
Traceback (most recent call last):
  File "/home/tom/.cache/pypoetry/virtualenvs/app-odcat-Dj0wURcA-py3.10/lib/python3.10/site-packages/uvicorn/protocols/http/httptools_impl.py", line 419, in run_asgi
    result = await app(  # type: ignore[func-returns-value]
  File "/home/tom/.cache/pypoetry/virtualenvs/app-odcat-Dj0wURcA-py3.10/lib/python3.10/site-packages/uvicorn/middleware/proxy_headers.py", line 78, in __call__
    return await self.app(scope, receive, send)
  File "/home/tom/project/odf/BlackSheep/blacksheep/server/application.py", line 801, in __call__
    return await self._handle_http(scope, receive, send)
  File "/home/tom/project/odf/BlackSheep/blacksheep/server/application.py", line 794, in _handle_http
    await send_asgi_response(response, send)
  File "blacksheep/scribe.pyx", line 327, in send_asgi_response
  File "/home/tom/.cache/pypoetry/virtualenvs/app-odcat-Dj0wURcA-py3.10/lib/python3.10/site-packages/uvicorn/protocols/http/httptools_impl.py", line 553, in send
    raise RuntimeError("Response content longer than Content-Length")
RuntimeError: Response content longer than Content-Length
```

**Reply by RobertoPrevato** [reply]
Thanks for investigating and all the insight. I will try to understand what is going on here.

**Reply by RobertoPrevato** [reply]
I tried uploading a 5.2 GB file using an HTML form with Firefox, to my proxy server in https://github.com/Neoteroi/BlackSheep-Examples/tree/main/proxy-2, with the fix applied locally (to use `long long` for `Content.length`) and could not reproduce the error `Invalid HTTP request received.` with uvicorn, even using `httptools` for it.

I will investigate more.

**Comment by thomafred** [comment]
I think the invalid http request message came from an issue with my own
implementation, and that I may have been generating more data than the
content length I provided to the StreamedContent constructor.

The fix alone seems to be enough :)
